ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -1019,4 +1031,173 @@ void markStartTime() {
+   *                              {@link Account#UNKNOWN_ACCOUNT} or {@link Container#UNKNOWN_CONTAINER}.
+   */
+  private void injectTargetAccountAndContainerFromBlobId(String blobIdStr, RestRequest restRequest)
+      throws RestServiceException {
+    BlobId blobId;
+    try {
+      blobId = new BlobId(blobIdStr, clusterMap);
+    } catch (Exception e) {
+      throw new RestServiceException("Invalid blob id=" + blobIdStr, RestServiceErrorCode.BadRequest);
+    }
+    Account targetAccount = accountService.getAccountById(blobId.getAccountId());
+    if (targetAccount == null) {
+      frontendMetrics.getHeadDeleteUnrecognizedAccountCount.inc();
+      // @todo The check can be removed once HelixAccountService is running with UNKNOWN_ACCOUNT created.
+      if (blobId.getAccountId() != Account.UNKNOWN_ACCOUNT_ID) {
+        throw new RestServiceException(
+            "Account from blobId=" + blobIdStr + "with accountId=" + blobId.getAccountId() + " cannot be recognized",
+            RestServiceErrorCode.InvalidAccount);
+      } else {
+        logger.debug(
+            "Account cannot be found for blobId={} with accountId={}. Setting targetAccount to UNKNOWN_ACCOUNT",
+            blobIdStr, blobId.getAccountId());
+        targetAccount = Account.UNKNOWN_ACCOUNT;
+      }
+    }
+    Container targetContainer = targetAccount.getContainerById(blobId.getContainerId());
+    if (targetContainer == null) {
+      frontendMetrics.getHeadDeleteUnrecognizedContainerCount.inc();
+      throw new RestServiceException(
+          "Container from blobId=" + blobIdStr + "with accountId=" + blobId.getAccountId() + " containerId="
+              + blobId.getContainerId() + " cannot be recognized", RestServiceErrorCode.InvalidContainer);
+    }
+    setTargetAccountAndContainerInRestRequest(restRequest, targetAccount, targetContainer);
+  }
+
+  /**
+   * Sanity check for {@link RestRequest}. This check ensures that the specified service id, account and container name,
+   * if they exist, should not be the same as the reserved values. It also makes sure certain headers must not be present.
+   * @param restRequest The {@link RestRequest} to check.
+   * @throws RestServiceException if the specified service id, account or container name is set as system reserved value.
+   */
+  private void accountAndContainerSanityCheck(RestRequest restRequest) throws RestServiceException {
+    if (Account.UNKNOWN_ACCOUNT_NAME.equals(getHeader(restRequest.getArgs(), Headers.TARGET_ACCOUNT_NAME, false))
+        || Account.UNKNOWN_ACCOUNT_NAME.equals(getHeader(restRequest.getArgs(), Headers.SERVICE_ID, false))) {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
+    }
+    String targetContainerName = getHeader(restRequest.getArgs(), Headers.TARGET_CONTAINER_NAME, false);
+    if (Container.UNKNOWN_CONTAINER_NAME.equals(targetContainerName) || Container.DEFAULT_PUBLIC_CONTAINER_NAME.equals(
```

Comment:
default public and private are not invalid containers - lets remove this check and add tests that check that blobs can be put into these containers.